R/iterate-responses.R

```diff
@@ -1,31 +1,63 @@
 #' Tools for working with lists of responses
 #'
-#' * `resps_combine()` combines the data from each response into a single
-#'   vector.
-#' * `resps_response()` returns all successful responses.
-#' * `resps_error()` returns all errors.
+#' @description
+#' These function provide a basic toolkit for operating with lists of
+#' responses as returned [req_perform_parallel()] and
+#' [req_perform_iteratively()].
+#'
+#' * `resps_successes()` returns a list successful responses.
+#' * `resps_failures()` returns a list failed responses (i.e. errors).
+#' * `resps_requests()` returns the list of requests that corresponds to
+#'   each request.
+#' * `resps_data()` returns all the data in a single vector or data frame.
+#'   It requires the vctrs package to be installed.
 #'
 #' @export
 #' @param resps A list of responses (possibly including errors).
 #' @param resp_data A function that takes a response (`resp`) and
 #'   returns the data foind inside that response as a vector or data frame.
-resps_combine <- function(resps, resp_data) {
-  check_installed("vctrs")
+#' @examples
+#' reqs <- list(
+#'   request(example_url()) |> req_url_path("/ip"),
+#'   request(example_url()) |> req_url_path("/user-agent"),
+#'   request(example_url()) |> req_template("/status/:status", status = 404),
+#'   request("INVALID")
+#' )
+#' resps <- req_perform_parallel(reqs)
+#'
+#' # find successful responses
+#' resps |> resps_successes()
+#'
+#' # collect all their data
+#' resps |> resps_successes() |> resps_data(\(resp) resp_body_json(resp))
+#'
+#' # find requests corresponding to failure responses
+#' resps |> resps_failures() |> resps_requests()
+resps_successes <- function(resps) {
+  resps[resps_ok(resps)]
+}
 
-  check_function2(resp_data, "resp")
-  vctrs::list_unchop(lapply(resps, resp_data))
+#' @export
+#' @rdname resps_successes
+resps_failures <- function(resps) {
+  resps[!resps_ok(resps)]
 }
-resps_is_resp <- function(resps) {
+
+resps_ok <- function(resps) {
   vapply(resps, is_response, logical(1))
 }
 
 #' @export
-#' @rdname resps_combine
-resps_responses <- function(resps) {
-  resps[resps_is_resp(resps)]
+#' @rdname resps_successes
+resps_requests <- function(resps) {
+  lapply(resps, function(x) x$request)
 }
+
 #' @export
-#' @rdname resps_combine
-resps_errors <- function(resps) {
-  resps[!resps_is_resp(resps)]
+#' @rdname resps_successes
+resps_data <- function(resps, resp_data) {
+  check_installed("vctrs")
+  check_function2(resp_data, "resp")
+
+  vctrs::list_unchop(lapply(resps, resp_data))
```

Comment:
Maybe `resps_data()` should check that all responses were successful. And I get this beauty when applying to an error:

``` r
devtools::load_all("~/GitHub/httr2/")
#> ℹ Loading httr2
reqs <- list(
  request(example_url()) |> req_template("/status/:status", status = 404),
  request("INVALID")
)
resps <- req_perform_parallel(reqs)

resps |> resps_data(\(resp) resp_body_json(resp))
#> Error in `env_has()` at httr2/R/resp-body.R:82:2:
#> ! `env` must be an environment, not `NULL`.
#> Backtrace:
#>     ▆
#>  1. └─httr2::resps_data(resps, function(resp) resp_body_json(resp))
#>  2.   ├─vctrs::list_unchop(lapply(resps, resp_data)) at httr2/R/iterate-responses.R:62:2
#>  3.   └─base::lapply(resps, resp_data)
#>  4.     └─global FUN(X[[i]], ...)
#>  5.       └─httr2::resp_body_json(resp)
#>  6.         └─rlang::env_has(env = resp$cache) at httr2/R/resp-body.R:82:2
#>  7.           └─rlang:::check_environment(env)
#>  8.             └─rlang:::stop_input_type(...)
#>  9.               └─rlang::abort(message, ..., call = call, arg = arg)
```

<sup>Created on 2023-10-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

This suggests that we should be more careful with the caching in `resp_body_json()` and `resp_body_xml()`

Reply:
Ooops, no, that's just because `check_response(resp)` should come before the cache lookup, not after 😬